### PR TITLE
Rewrite `OverflowSafeInt` to avoid undefined behaviour

### LINKED ISSUE #8284: Negating or taking absolute values of maximally-negative OverflowSafeInt may result in undefined behaviour
As mentioned in https://github.com/OpenTTD/OpenTTD/pull/7924, `OverflowSafeInt` can produce incorrect results when the value underflows. This appears to be the result of undefined behaviour caused by taking the inverse or absolute value of T_MIN; the magnitude of which cannot usually be represented as a positive value of the same integer type.

I am not yet familiar with OpenTTD's continuous integration and regression testing system; however, I have an idea to work around the issue using typecasts to unsigned integers where appropriate.

### PR DESCRIPTION
**Pull request by techgeeknz**
The current implementation of `OverflowSafeInt` does not handle the maximally-negative boundary condition very well. This rewrite uses an unsigned type to calculate and represent absolute values, to avoid invoking undefined behaviours such as `abs(INT64_MIN) < 0` and `-INT64_MIN == INT64_MIN`. Fixes #8284.

### CODE REVIEW

**Comment by nielsmh** [comment]
I don't think the regression test suite exercises OverflowSafeInt as it is. I can't think of a good way to do that without major changes to the test, but it's probably worth setting as a goal.

**Comment by techgeeknz** [comment]
I'm willing to write tests for it, if somebody can point me in the direction of the documentation for the test suite. Presumably, these make use of the squirrel API in some shape or form?

**Comment by nielsmh** [comment]
Is there any documentation for the tests at all? It works by loading a specific savegame with an AI script, the AI script then does various actions and writes various values to logging. After executing for a number of game ticks the game exits, and the console output is compared to an expected output.

So there's three parts to potentially modify: The AI script to add/change behaviours to exercise. The expected output if the expected behaviour changes. The savegame if some basic premises need to be changed. (Changing the savegame is probably rather risky.)

**Comment by techgeeknz** [comment]
Nope. I'm not touching that with a 10-foot barge pole 😎.

**Comment by LordAro** [comment]
Yeah, it would be difficult to get close to over/underflow testing with the AI scripts anyway - certainly extremely time consuming (talking hours) and not something that's feasible for regression tests

**Comment by techgeeknz** [comment]
I guess the code is nowhere near modular enough for a TDD design with dependency injection to ever be feasible?

**Comment by techgeeknz** [comment]
With https://github.com/OpenTTD/OpenTTD/pull/8285/commits/e97dd52ba33dbac952d050c5304ed0a055d44de9, the only instances of `OverflowSafeInt` are via the `OverflowSafeInt32` and `OverflowSafeInt64` typedefs. Would there be any objection to simplifying the template to use `std::numeric_limits<T>` instead of (the impression of) freely-selectable saturation limits?
My concern is that the current implementation assumes that `T_MIN` is negative and `T_MAX` is positive, and that both are of similar magnitude (whereas the previous implementation assumed that `T_MIN == -T_MAX`); and with user-selected limits, that assumption may be invalid. At the very least, we could enforce that the type parameter is a signed integer type.

**Comment by JGRennison** [comment]
Ideally this check would just be a static_assert, but that is probably asking a bit much when targeting C++11.

A command line switch which ran some unit-style tests, output any required log messages, and then exited with the appropriate exit code would be a simple/automatable way to test things like this which don't require a full game state.

**Comment by LordAro** [comment]
Go for it, don't see why not

**Comment by LordAro** [comment]
Can't say I'm a fan of a "command line switch" solution to doing unit tests - I don't really want test code in the actual "game", as it were.

A separate executable that actually runs/tests the non-side-effecty bits that you actually can run unit tests on should be the preferred solution.

**Comment by techgeeknz** [comment]
> A separate executable that actually runs/tests the non-side-effecty bits that you actually can run unit tests on should be the preferred solution.

If said executable could (optionally) be built and run during the compilation process; that would be even better :smile:

**Comment by LordAro** [comment]
For really simple stuff, static_assert/assert_compile should be preferred, of course

**Comment by techgeeknz** [comment]
Hmm, I don't know what I've done to upset `kdevelop`, but it seems to be crashing quite a lot lately. I may have to hunt for a better code editor :roll_eyes:.

File: src/core/overflowsafe_type.hpp [hunk 1]
```diff
@@ -72,7 +78,12 @@ class OverflowSafeInt
 	OverflowSafeInt() : m_value(0) { }
 
 	OverflowSafeInt(const OverflowSafeInt& other) { this->m_value = other.m_value; }
-	OverflowSafeInt(const int64 int_)             { this->m_value = int_; }
+	OverflowSafeInt(const intmax_t int_)
+	{
+		this->m_value = unlikely(int_ > static_cast<intmax_t>(T_MAX)) ? T_MAX :
+				unlikely(int_ < static_cast<intmax_t>(T_MIN)) ? T_MIN :
+				static_cast<T>(int_);
+	}
```
Comment (techgeeknz) [inline src/core/overflowsafe_type.hpp]: I would like to have a conversion constructor for `uintmax_t`; unfortunately, the compiler didn't seem to be able to resolve the ambiguity when offered the choice between `intmax_t` and `uintmax_t`.

File: src/core/overflowsafe_type.hpp [hunk 2]
```diff
@@ -134,72 +150,81 @@ class OverflowSafeInt
...
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value && (static_cast<intmax_t>(std::numeric_limits<I>::min()) >= static_cast<intmax_t>(T_MIN)) && (static_cast<uintmax_t>(std::numeric_limits<I>::max()) <= static_cast<uintmax_t>(T_MAX))>::type>
+	inline OverflowSafeInt& operator /= (const I                divisor)       { this->m_value /= static_cast<T>(divisor); return *this; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value && (static_cast<intmax_t>(std::numeric_limits<I>::min()) >= static_cast<intmax_t>(T_MIN)) && (static_cast<uintmax_t>(std::numeric_limits<I>::max()) <= static_cast<uintmax_t>(T_MAX))>::type>
+	inline OverflowSafeInt  operator /  (const I                divisor) const { OverflowSafeInt result = *this; result /= static_cast<T>(divisor); return result; }
+	inline OverflowSafeInt& operator /= (const intmax_t         divisor)       { *this = *this / divisor; return *this; }
+	inline OverflowSafeInt  operator /  (const intmax_t         divisor) const { return OverflowSafeInt(static_cast<intmax_t>(this->m_value) / static_cast<intmax_t>(divisor)); }
```
Comment (techgeeknz) [inline src/core/overflowsafe_type.hpp]: I am not 100% sure this construct will work as intended. The code does at least compile, though.

File: src/core/overflowsafe_type.hpp [hunk 3]
```diff
@@ -104,25 +115,30 @@ class OverflowSafeInt
 
 	/* Operators for addition and subtraction */
 	inline OverflowSafeInt  operator +  (const OverflowSafeInt& other) const { OverflowSafeInt result = *this; result += other; return result; }
-	inline OverflowSafeInt  operator +  (const int              other) const { OverflowSafeInt result = *this; result += (int64)other; return result; }
-	inline OverflowSafeInt  operator +  (const uint             other) const { OverflowSafeInt result = *this; result += (int64)other; return result; }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+	inline OverflowSafeInt& operator += (const I                other)       { return *this += OverflowSafeInt(other); }
+	template <class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+	inline OverflowSafeInt  operator +  (const I                other) const { return *this +  OverflowSafeInt(other); }
```
Comment (techgeeknz) [inline src/core/overflowsafe_type.hpp]: (I think) we need to typecast, in order to be sure `other` is of a compatible type to `this->m_value`.

File: src/core/overflowsafe_type.hpp [hunk 4]
```diff
@@ -134,72 +150,81 @@ class OverflowSafeInt
...
 
 	/* Operators for shifting */
-	inline OverflowSafeInt& operator <<= (const int shift)       { this->m_value <<= shift; return *this; }
-	inline OverflowSafeInt  operator <<  (const int shift) const { OverflowSafeInt result = *this; result <<= shift; return result; }
-	inline OverflowSafeInt& operator >>= (const int shift)       { this->m_value >>= shift; return *this; }
-	inline OverflowSafeInt  operator >>  (const int shift) const { OverflowSafeInt result = *this; result >>= shift; return result; }
+	inline OverflowSafeInt  operator <<  (const int shift) const { return OverflowSafeInt(this->m_value << shift); }
+	inline OverflowSafeInt& operator <<= (const int shift)       { *this = *this << shift; return *this; }
+	inline OverflowSafeInt  operator >>  (const int shift) const { return OverflowSafeInt(this->m_value >> shift); }
+	inline OverflowSafeInt& operator >>= (const int shift)       { *this = *this >> shift; return *this; }
```
Comment (techgeeknz) [inline src/core/overflowsafe_type.hpp]: Typecast is supposed to ensure that the value is not shifted outside the valid range; however, I just realised this won't actually work :wink:.

File: src/core/overflowsafe_type.hpp [hunk 5]
```diff
@@ -134,72 +150,81 @@ class OverflowSafeInt
...
-typedef OverflowSafeInt<int32, INT32_MAX, INT32_MIN> OverflowSafeInt32;
+/* Sometimes we got integral operator OverflowSafeInt instead of vice versa. Handle that properly */
+template <class T, class U, class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+inline OverflowSafeInt<T, U> operator + (I a, OverflowSafeInt<T, U> b) { return b + a; }
+template <class T, class U, class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+inline OverflowSafeInt<T, U> operator - (I a, OverflowSafeInt<T, U> b) { return -b + a; }
+template <class T, class U, class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+inline OverflowSafeInt<T, U> operator * (I a, OverflowSafeInt<T, U> b) { return b * a; }
+template <class T, class U, class I, class = typename std::enable_if<std::is_integral<I>::value>::type>
+inline OverflowSafeInt<T, U> operator / (I a, OverflowSafeInt<T, U> b) { return OverflowSafeInt<T, U>(a) / b; }
```
Comment (techgeeknz) [inline src/core/overflowsafe_type.hpp]: Might as well make it work for *all* integral types :smile:.

**Comment by James103** [comment]
> Yeah, it would be difficult to get close to over/underflow testing with the AI scripts anyway - certainly extremely time consuming (talking hours) and not something that's feasible for regression tests

Could a specific regression test (with different savegame and AI scripts) be prepared just to test for those boundary conditions, where the AI company's money is hacked to be near INT64_MIN?

**Comment by techgeeknz** [comment]
I have just started a new job, so I am not able to devote as much time to this as I would like to. If anybody else wants to clone the branch and finish the work, please feel free to do so.

**Comment by techgeeknz** [comment]
There are two issues that I cannot seem to resolve:

1. Why does the regression test fail with what looks like a bunch of off-by-one errors when I implement a templated constructor that accepts any integral type?
2. Just what exactly is MSVC's problem? Why does it choke so badly on this?

**Comment by LordAro** [comment]
I feel like the changes in this PR have gone a bit "out of scope". Can you explain the `RawValue` additions? Why is `operator()` not sufficient?

**Comment by techgeeknz** [comment]
> I feel like the changes in this PR have gone a bit "out of scope". Can you explain the `RawValue` additions? Why is `operator()` not sufficient?

Good question. By requiring explicit typecasting from `OverflowSafeInt` to regular overflow-unsafe integrals, I was able to uncover a number of bugs and potential problems that were introduced by unintentional implicit typecasts to overflow-unsafe integers.

I had originally intended to use `explicit operator T ()` for this purpose, but it soon became evident that the type to be cast is often not readily apparent from the context in which it is used. Perhaps a better method name will suggest itself?

Having said that, a templated explicit typecast to integral might be plausible; as it seems reasonable that, in situations where it makes sense to cast away from `OverflowSafeInt`, the data type required for the correct result should be obvious.

File: src/graph_gui.cpp [hunk 6]
```diff
@@ -223,8 +223,8 @@ struct BaseGraphWindow : Window {
 		current_interval.lowest =  (11 * current_interval.lowest) / 10;
 
 		/* Always include zero in the shown range. */
-		double abs_lower  = (current_interval.lowest > 0) ? 0 : (double)abs(current_interval.lowest);
-		double abs_higher = (current_interval.highest < 0) ? 0 : (double)current_interval.highest;
+		double abs_lower  = (current_interval.lowest > 0) ? 0 : abs(current_interval.lowest);
+		double abs_higher = (current_interval.highest < 0) ? 0 : abs(current_interval.highest);
```
Comment (LordAro) [inline src/graph_gui.cpp:227]: Instead of adding a redundant abs here, might be better to change the preceding line to `-current_interval.lowest` ?

File: src/network/network_server.cpp [hunk 7]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (LordAro) [inline src/network/network_server.cpp]: :(

File: src/script/squirrel_helper.hpp [hunk 8]
```diff
@@ -15,6 +15,7 @@
 #include "../economy_type.h"
 #include "../string_func.h"
 #include "squirrel_helper_type.hpp"
+#include <3rdparty/squirrel/include/squirrel.h>
```
Comment (LordAro) [inline src/script/squirrel_helper.hpp]: Why is this necessary?

File: src/graph_gui.cpp (hunk 6 above)
Comment (techgeeknz) [reply src/graph_gui.cpp:227]: The `abs` isn't actually redundant. `abs(OverflowSafeInt)` always returns an unsigned integer, so that `abs(T_MIN)` will work correctly.

File: src/script/squirrel_helper.hpp (hunk 8 above)
Comment (LordAro) [inline src/script/squirrel_helper.hpp]: Regardless, should be `#include <squirrel.h>` (3rdparty/squirrel/include is on the searchpath)

File: src/network/network_server.cpp (hunk 7 above)
Comment (techgeeknz) [reply src/network/network_server.cpp]: Ah, so that's where that wayward carriage return ended up :thinking:

File: src/graph_gui.cpp (hunk 6 above)
Comment (LordAro) [reply src/graph_gui.cpp:227]: Probably wants a comment to that effect then :)

File: src/script/squirrel_helper.hpp (hunk 8 above)
Comment (techgeeknz) [reply src/script/squirrel_helper.hpp]: It isn't; but `kdevelop` insisted that it was. I'll remove it.

File: src/graph_gui.cpp (hunk 6 above)
Comment (techgeeknz) [reply src/graph_gui.cpp:227]: Good idea.

**Comment by techgeeknz** [comment]
Now that the program actually compiles under MSVC _and_ the regression tests pass, I'm more than happy to address any concerns you may have with the `RawValue()` method and any other code quality issues.

**Comment by techgeeknz** [comment]
> Having said that, a templated explicit typecast to integral might be plausible; as it seems reasonable that, in situations where it makes sense to cast away from `OverflowSafeInt`, the data type required for the correct result should be obvious.

Hmm, I'm not convinced this is better. What are your thoughts, @LordAro?

**Comment by techgeeknz** [comment]
@LordAro would you be more comfortable if I were to separate the bugfixes for the implicit typecasts into a supplimentary pull request; given that it's something that probably should be fixed but leaving them out won't make anything more broken than it already is?

**Comment by James103** [comment]
Is this PR still relevant? The money underflow exploit described in #8284 is fixed in JGRPP 0.42.0 but not in master as of commit 513641f.
